src/core/particle_node.cpp

```diff
@@ -231,29 +231,21 @@ static std::vector<Particle> mpi_get_particles(std::span<const int> ids) {
 
   for (auto const &p_id : ids) {
     auto const p_node = get_particle_node(p_id);
-
-    if (p_node != this_node)
-      node_ids[p_node].push_back(p_id);
+    node_ids[p_node].emplace_back(p_id);
   }
+  /* We shouldn't be prefetching particles that are already on the head node */
+  assert(node_ids[this_node].empty());
 
-  /* Distributed ids to the nodes */
+  /* Distribute ids to the worker nodes */
   {
     static std::vector<int> ignore;
     boost::mpi::scatter(comm_cart, node_ids, ignore, 0);
+    assert(ignore.empty());
   }
 
-  /* Copy local particles */
-  std::transform(node_ids[this_node].cbegin(), node_ids[this_node].cend(),
-                 parts.begin(), [](int p_id) {
-                   auto const p = get_cell_structure().get_local_particle(p_id);
-                   assert(p != nullptr);
-                   return *p;
-                 });
```

Comment:
Does this mean that we don't have particles of the head node in the cache after the call to `prefetch_particle_data`?

Reply:
To my understanding, the cache only holds copies of particles from worker nodes, because particles on the head node are always immediately accessible. In fact, there is a pre-filtering of the user input to remove particles who have a valid pointer on rank 0, so this transform was always a no-op. In fact, when a simulation only has 1 MPI rank, the function is never called:
https://github.com/espressomd/espresso/blob/adf636d9916a19485604c376d925d7849fba3076/src/core/particle_node.cpp#L264-L268